src/main/scala/scalaz/ioeffect/IORef.scala

```diff
@@ -18,34 +18,108 @@ import java.util.concurrent.atomic.AtomicReference
  * } yield ()
  * }}}
  */
-final class IORef[A] private (private val value: AtomicReference[A]) {
+final class IORef[A] private (private val value: AtomicReference[A])
+    extends AnyVal {
 
   /**
    * Reads the value from the `IORef`.
    */
-  final def read: IO[A] = IO.sync(value.get())
+  final def read[E]: IO[E, A] = IO.sync(value.get)
+
+  /**
+   * Writes a new value to the `IORef`, with a guarantee of immediate
+   * consistency (at some cost to performance).
+   */
+  final def write[E](a: A): IO[E, Unit] = IO.sync(value.set(a))
 
   /**
-   * Writes a new value to the `IORef`.
+   * Writes a new value to the `IORef` without providing a guarantee of
+   * immediate consistency.
    */
-  final def write(a: A): IO[Unit] = IO.sync(value.set(a))
+  final def writeLater[E](a: A): IO[E, Unit] = IO.sync(value.lazySet(a))
 
   /**
-   * Modifies the `IORef` with the specified function.
+   * Attempts to write a new value to the `IORef`, but aborts immediately under
+   * concurrent modification of the value by other fibers.
    */
-  final def modify(f: A => A): IO[A] = IO.sync {
-    val cur  = value.get()
-    val next = f(cur)
-    value.set(next)
+  final def tryWrite[E](a: A): IO[E, Boolean] =
```

Comment:
great stuff here! @jdegoes OOI, what is the point of this method? it only fails under high write contention where the read value is immediately changed. There is nothing deterministic here. It would seem to me that supplying the old value would make more sense, accessing the compare&set semantics from the client, eg:

    def tryWrite[E](old: A, a: A): IO[E, Boolean] =
       IO.sync(value.compareAndSet(old, a))

Reply:
I can't think of a use-case for `tryWrite`, but I know of many for `compareAndSet` - it says, write X to Y as long as X is still _the same_ X. Reference equality is important as it disallows the ABA problem, where someone changes A to B, then another change turns that back to an equivalent A. The field has changed, so you don't want CAS to work. This is what makes lock-free programming much harder in C.

The notion that `tryWrite` gives you feedback of high-contention is odd and doesn't seem useful, what it gives you is an indication that some write accidentally failed due to factors totally outside your control. What are you meant to do with this? Well, you'd generally just try again and it will likely work. I cannot imagine how I'd use this as an IORef client.

Reply:
Well, you can use `tryWrite` in a loop to build your own `modify` that takes different actions under contention. That said, I think you've made a good case for a low-level CAS. I'll add one!